lsp/nls/src/requests/completion.rs

```diff
@@ -21,42 +29,194 @@ pub fn handle_completion(
+                contracts.iter().find_map(|contract| {
+                    if let Types(AbsType::Record(..)) = contract.types {
+                        Some(contract.clone())
+                    } else {
+                        None
+                    }
+                })
+            }
+            _ if item.id == id => match item.kind {
+                TermKind::Declaration(_, _, ValueState::Known(new_id))
+                | TermKind::Usage(UsageState::Resolved(new_id)) => {
+                    find_record_contract(&linearization, new_id)
+                }
+                _ => None,
+            },
+            _ => None,
+        })
+    }
 
+    /// Extract identifiers from a row type which forms a record.
+    /// Ensure that ty is really a row type.
+    fn extract_ident(ty: &Box<Types>) -> Vec<Ident> {
+        ty.iter_as_rows()
+            .filter_map(|item| match item {
+                RowIteratorItem::Row(ident, _) => Some(ident.clone()),
+                RowIteratorItem::Tail(..) => None,
+            })
+            .collect::<Vec<_>>()
+    }
+
+    let item = linearization.item_at(&locator);
     if item == None {
+        // NOTE: this might result in a case where completion is not working.
+        // This happens when an item cannot be found from the linearization.
         server.reply(Response::new_ok(id, Value::Null));
         return Ok(());
     }
-
     let item = item.unwrap().to_owned();
 
-    let in_scope: Vec<_> = linearization
-        .get_in_scope(&item)
+    Trace::enrich(&id, linearization);
+
+    let trigger = params.context.and_then(|context| context.trigger_character);
+    let in_scope = match trigger {
+        // Record Completion
+        Some(s) if s == "." => {
+            // Get the ID before the .
+            // if the current source is `let var = {a = 10} in var.`
+            // name = "var"
+            let name = text[..start - 1]
+                .chars()
+                .rev()
+                .skip_while(|c| *c == '.') // skip . (if any)
+                .take_while(|c| c.is_ascii_alphabetic())
+                .collect::<String>()
+                .chars()
+                .rev()
+                .collect::<String>();
+
+            let name_id = lin_env
+                .iter()
+                .find(|(ident, _)| ident.label() == name)
+                .map(|(_, id)| *id);
+            if name_id == None {
+                // NOTE: this might result in a case where completion is not working.
+                // This happens when an item cannot be found from the linearization.
+                server.reply(Response::new_ok(id, Value::Null));
+                return Ok(());
+            }
+            let name_id = name_id.unwrap();
+
+            linearization
+                .get_in_scope(&item)
+                .iter()
+                .filter_map(|i| {
+                    let (ty, _) = linearization.resolve_item_type_meta(i);
```

Comment:
I would say using `_` is common in pattern matching and destructuring in Rust (and other functional languages). As a datapoint, I personally do that in OCaml and Rust, because somehow if you don't use this variable, it should be irrelevant to understanding the code (if the thing you extract is correctly named), same as using ellipsis `..` in Rust when matching on a big struct:

```rust
if let Foo { bar: "bar", baz, ..} = value {
 ..
}
```

If we follow your convention, and `Foo` has 10 fields, we should know and write them all, which I don't personally find that useful. But, as with many things, it's also a matter of taste :) I think both approach are used and reasonable.

Reply:
While I won't push further on this, I think the ellipsis is a different topic.

Tuple fields are unnamed: there is no trivial way to distinguish the parts other than the order, which is brittle. Even worse when the different tuple fields are of the same type. A change of order of fields in the type and return of the function returning a tuple could have dramatic, undetected effects.

Ellipsis happens on named fields (to the best of my knowledge) so the risk is significantly lower: unless one changes both the fields of a record AND also swaps field names (which is an extremely dangerous thing to do :smile: ), an ellipsis can't break the code at runtime in the future.

Reply:
You're absolutely right about the advantage of named fields vs unnamed fields. I think it does speak for using them as much as possible when defining datatypes with more than one argument (which is something we haven't always done in the past). For example, you could argue that giving a proper type to the result of `resolve_item_type_meta` is a good idea.

However, once we have to deal with a tuple, as the caller, I fail to see how giving a name to an unused variable protects against any risk you mention. Program A with `_` and program A' with `_named` are strictly equivalent with respect to runtime behavior and compilation. So if you refactor the order of fields, your program will either silently compile and do the wrong thing at runtime, or break at compile time in the same way, whether you give names to unused fields or not.

In the end, naming an unused field is merely adding a non-checked comment for readers. Which is arguably not a bad thing to do, but I just wanted to give another perspective, as I personally prefer not to add mental load for something that can be ignored, and to say that it's more a matter of taste. I trust Gaga to make an informed choice now :slightly_smiling_face: